server/aws-lsp-codewhisperer/src/language-server/agenticChat/tools/fsReplace.ts

```diff
@@ -0,0 +1,159 @@
+                '- When multiple edits to the same file are needed, ALWAYS populate the diffs array with `oldStr` and `newStr` pairs. This improves efficiency by reducing the number of tool calls and ensures the file remains in a consistent state',
+            inputSchema: {
+                type: 'object',
+                properties: {
+                    explanation: {
+                        description:
+                            'One sentence explanation as to why this tool is being used, and how it contributes to the goal.',
+                        type: 'string',
+                    },
+                    diffs: {
+                        description: 'List of diffs to replace contents in an existing file.',
+                        type: 'array',
+                        items: {
+                            type: 'object',
+                            properties: {
+                                oldStr: {
+                                    description:
+                                        'The exact string content to be replaced in the file. Must match exactly including whitespace and line breaks.',
+                                    type: 'string',
+                                },
+                                newStr: {
+                                    description:
+                                        'The new string content that will replace the oldStr. Use empty string to delete content.',
+                                    type: 'string',
+                                },
+                            },
+                            required: ['oldStr'],
+                        },
+                    },
+                    path: {
+                        description:
+                            'Absolute path to a file, e.g. `/repo/file.py` for Unix-like system including Unix/Linux/macOS or `d:\\repo\\file.py` for Windows.',
+                        type: 'string',
+                    },
+                },
+                required: ['diffs', 'path'],
+            },
+        } as const
+    }
+}
+
+const getReplaceContent = (params: ReplaceParams, fileContent: string) => {
+    // Detect line ending from oldContent (CRLF, LF, or CR)
+    const match = fileContent.match(/\r\n|\r|\n/)
+    const lineEnding = match ? match[0] : os.EOL
+
+    for (const diff of params.diffs) {
+        if (diff.newStr == undefined) {
+            diff.newStr = ''
+        }
+        // Normalize oldStr and newStr to match fileContent's line ending style
+        const normalizedOldStr = diff.oldStr.split(/\r\n|\r|\n/).join(lineEnding)
+        const normalizedNewStr = diff.newStr.split(/\r\n|\r|\n/).join(lineEnding)
+
+        const matches = [...fileContent.matchAll(new RegExp(escapeRegExp(normalizedOldStr), 'g'))]
+
+        if (matches.length === 0) {
+            throw new Error(`No occurrences of "${diff.oldStr}" were found`)
```

Comment:
Should we return "X/Y diffs were applied before this failure" to help with partial failures?

Reply:
good callout, I will update the tool

Reply:
updated